File: Src/java/cql-to-elm/src/jvmTest/java/org/cqframework/cql/cql2elm/TestLibrarySourceProvider.java
package org.cqframework.cql.cql2elm;

import kotlinx.io.Source;
import org.hl7.elm.r1.VersionedIdentifier;

import static kotlinx.io.CoreKt.buffered;
import static kotlinx.io.JvmCoreKt.asSource;

public class TestLibrarySourceProvider implements LibrarySourceProvider {

    private String path = "LibraryTests";

    public TestLibrarySourceProvider() {}

    public TestLibrarySourceProvider(String path) {
        this.path = path;
    }

    @Override
    public Source getLibrarySource(VersionedIdentifier libraryIdentifier) {
        return getLibraryContent(libraryIdentifier, LibraryContentType.CQL);
    }

    @Override
    public Source getLibraryContent(VersionedIdentifier libraryIdentifier, LibraryContentType type) {
        var inputStream = TestLibrarySourceProvider.class.getResourceAsStream(getFileName(libraryIdentifier, type));
        return inputStream == null ? null : buffered(asSource(inputStream));
    }

    private String getFileName(VersionedIdentifier libraryIdentifier, LibraryContentType type) {
        return String.format(
                "%s/%s%s.%s",
                path,
                libraryIdentifier.getId(),
                libraryIdentifier.getVersion() != null ? ("-" + libraryIdentifier.getVersion()) : "",
                type.toString().toLowerCase());
    }
}

File: Src/java/cql-to-elm/src/jvmTest/kotlin/org/cqframework/cql/cql2elm/TestLibrarySourceProvider.kt
package org.cqframework.cql.cql2elm

import kotlinx.io.Source
import kotlinx.io.asSource
import kotlinx.io.buffered
import org.hl7.elm.r1.VersionedIdentifier

class TestLibrarySourceProvider(val path: String? = "LibraryTests") : LibrarySourceProvider {

    override fun getLibrarySource(libraryIdentifier: VersionedIdentifier): Source? {
        return getLibraryContent(libraryIdentifier, LibraryContentType.CQL)
    }

    override fun getLibraryContent(
        libraryIdentifier: VersionedIdentifier,
        type: LibraryContentType,
    ): Source? {
        val inputStream =
            TestLibrarySourceProvider::class
                .java
                .getResourceAsStream(getFileName(libraryIdentifier, type))
        return inputStream?.asSource()?.buffered()
    }

    private fun getFileName(
        libraryIdentifier: VersionedIdentifier,
        type: LibraryContentType,
    ): String {
        val version = libraryIdentifier.version?.let { "-$it" } ?: ""
        return "$path/${libraryIdentifier.id}${version}.${type.toString().lowercase()}"
    }
}
